cpp/dolfinx/fem/interpolate.h

```diff
@@ -952,7 +936,70 @@ void interpolate(Function<T>& u, const Function<T>& v,
   {
     // Get mesh and check that functions share the same mesh
     if (auto mesh_v = v.function_space()->mesh(); mesh != mesh_v)
-      impl::interpolate_nonmatching_meshes(u, v, cells);
+    {
+      int result;
+      MPI_Comm_compare(mesh->comm(), mesh_v->comm(), &result);
+
+      if (result == MPI_UNEQUAL)
+        throw std::runtime_error("Interpolation on different meshes is only "
+                                 "supported with the same communicator.");
+
+      MPI_Comm comm = mesh->comm();
+      const int tdim = mesh->topology().dim();
+      const auto cell_map = mesh->topology().index_map(tdim);
+
+      std::shared_ptr<const FiniteElement> element_u
+          = u.function_space()->element();
+      const std::size_t value_size = element_u->value_size();
+
+      std::tuple_element_t<0, M2MInterpolationData> dest_ranks;
+      std::tuple_element_t<1, M2MInterpolationData> src_ranks;
+      std::tuple_element_t<2, M2MInterpolationData> received_points;
+      std::tuple_element_t<3, M2MInterpolationData> evaluation_cells;
+
+      if (m2mInterpolationData == nullptr)
```

Comment:
Best to make the expensive call mandatory, users then are 'forced' to understand that it is there.